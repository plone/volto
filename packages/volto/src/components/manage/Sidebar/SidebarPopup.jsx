import React from 'react';
import { createPortal } from 'react-dom';
import { CSSTransition } from 'react-transition-group';
import PropTypes from 'prop-types';
import { doesNodeContainClick } from 'semantic-ui-react/dist/commonjs/lib';

const DEFAULT_TIMEOUT = 500;

const SidebarPopup = (props) => {
  const { children, open, onClose, overlay } = props;

  const asideElement = React.useRef();

  const handleClickOutside = (e) => {
    if (asideElement && doesNodeContainClick(asideElement.current, e)) return;
    onClose();
  };

  React.useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside, false);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside, false);
    };
  });

  return (
    <>
      {overlay && (
        <CSSTransition
          in={open}
          timeout={DEFAULT_TIMEOUT}
          classNames="overlay-container"
          unmountOnExit
        >
          <>
            {document?.body &&
              createPortal(
                <div className="overlay-container"></div>,
                document?.body,
              )}
          </>
        </CSSTransition>
      )}
      <CSSTransition
        in={open}
        timeout={DEFAULT_TIMEOUT}
        classNames="sidebar-container"
        unmountOnExit
      >
        <>
          {createPortal(
            <aside
              role="presentation"
              onClick={(e) => {
                e.stopPropagation();
              }}
              onKeyDown={(e) => {
                e.stopPropagation();
              }}
              ref={asideElement}
              key="sidebarpopup"
              className="sidebar-container"
              style={{ overflowY: 'auto' }}
            >
              {children}
            </aside>,
            document.body,
          )}
        </>
      </CSSTransition>
    </>
  );
};

SidebarPopup.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  overlay: PropTypes.bool,
};

SidebarPopup.defaultProps = {
  open: false,
  onClose: () => {},
  overlay: false,
};

export default SidebarPopup;
