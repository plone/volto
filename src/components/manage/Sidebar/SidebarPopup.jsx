import React from 'react';
import { Portal } from 'react-portal';
import { CSSTransition } from 'react-transition-group';
import PropTypes from 'prop-types';
import { doesNodeContainClick } from 'semantic-ui-react/dist/commonjs/lib';

const DEFAULT_TIMEOUT = 500;

const SidebarPopup = (props) => {
  const { children, open, onClose, overlay } = props;

  const asideElement = React.createRef();

  const handleClickOutside = (e) => {
    if (asideElement && doesNodeContainClick(asideElement.current, e)) return;
    onClose();
  };

  React.useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside, false);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside, false);
    };
  });

  return (
    <>
      {overlay && (
        <CSSTransition
          in={open}
          timeout={DEFAULT_TIMEOUT}
          classNames="overlay-container"
          unmountOnExit
        >
          <Portal node={document?.body}>
            <div className="overlay-container"></div>
          </Portal>
        </CSSTransition>
      )}
      <CSSTransition
        in={open}
        timeout={DEFAULT_TIMEOUT}
        classNames="sidebar-container"
        unmountOnExit
      >
        <Portal>
          <aside
            role="presentation"
            onClick={(e) => {
              e.stopPropagation();
            }}
            onKeyDown={(e) => {
              e.stopPropagation();
            }}
            ref={asideElement}
            key="sidebarpopup"
            className="sidebar-container"
            style={{ overflowY: 'auto' }}
          >
            {children}
          </aside>
        </Portal>
      </CSSTransition>
    </>
  );
};

SidebarPopup.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  overlay: PropTypes.bool,
};

SidebarPopup.defaultProps = {
  open: false,
  onClose: () => {},
  overlay: false,
};

export default SidebarPopup;
