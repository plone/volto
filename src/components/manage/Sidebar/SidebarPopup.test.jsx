import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import SidebarPopup from './SidebarPopup';

test('sidebar popup is rendered when the block is selected', () => {
  const renderer = new ShallowRenderer();
  renderer.render(
    <SidebarPopup open={true}>
      <p>Tested!</p>
    </SidebarPopup>,
  );
  const component = renderer.getRenderOutput();
  expect(component).toMatchSnapshot();
});

test('sidebar popup is not rendered when the block is not selected', () => {
  const renderer = new ShallowRenderer();
  renderer.render(
    <SidebarPopup open={false}>
      <p>Tested, but you shouldn't see this in the snapshot!</p>
    </SidebarPopup>,
  );
  const component = renderer.getRenderOutput();
  expect(component).toMatchSnapshot();
});
