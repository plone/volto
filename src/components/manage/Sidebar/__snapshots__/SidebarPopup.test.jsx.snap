// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar popup is not rendered when the block is not selected 1`] = `
<React.Fragment>
  <CSSTransition
    classNames="sidebar-container"
    in={false}
    timeout={500}
    unmountOnExit={true}
  >
    <Portal>
      <aside
        className="sidebar-container"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        style={
          Object {
            "overflowY": "auto",
          }
        }
      >
        <p>
          Tested, but you shouldn't see this in the snapshot!
        </p>
      </aside>
    </Portal>
  </CSSTransition>
</React.Fragment>
`;

exports[`sidebar popup is rendered when the block is selected 1`] = `
<React.Fragment>
  <CSSTransition
    classNames="sidebar-container"
    in={true}
    timeout={500}
    unmountOnExit={true}
  >
    <Portal>
      <aside
        className="sidebar-container"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        style={
          Object {
            "overflowY": "auto",
          }
        }
      >
        <p>
          Tested!
        </p>
      </aside>
    </Portal>
  </CSSTransition>
</React.Fragment>
`;
